src/tiled/mapdocument.h

```diff
@@ -24,6 +24,7 @@
 
 #include "document.h"
 #include "layer.h"
+#include "mapformat.h"
```

Comment:
I believe you shouldn't need this include and can keep relying on the forward declaration.

Reply:
Ahh this one is actually legit. I've modified ``exportFormat`` to have a covariant return type, so that derived classes can return a pointer that is descended from ``FileFormat`` e.g. ``MapFormat`` etc which is a little nicer. Because of this the [full definition of the class needs to exist](https://stackoverflow.com/questions/29817748/invalid-covariant-return-type-error) and [a forward declaration is inadequate](https://stackoverflow.com/questions/2159390/in-c-is-it-possible-to-forward-declare-a-class-as-inheriting-from-another-cla), so that the compiler knows that ``MapFormat`` is derived from ``FileFormat`` and the covariant return is valid.

Reply:
Good point (and what an annoying top-voted answers in your second SO link...). :-)

Reply:
Haha yeah, at least the top voted comment for that answer pulls it back into line a bit.